src/read.rs

```diff
@@ -181,6 +197,34 @@ impl<'a> ZipFileReader<'a> {
             ZipFileReader::NoReader => panic!("ZipFileReader was in an invalid state"),
             ZipFileReader::Raw(r) => r,
             ZipFileReader::Stored(r) => r.into_inner().into_inner(),
+            #[cfg(feature = "legacy-zip")]
+            ZipFileReader::Shrink(r) => {
+                // Lzma reader owns its buffer rather than mutably borrowing it, so we have to drop
+                // it separately
+                if let Ok(mut remaining) = r.into_inner().finish() {
+                    let _ = copy(&mut remaining, &mut sink());
+                }
+                return;
+            }
+            #[cfg(feature = "legacy-zip")]
+            ZipFileReader::Reduce(r) => {
+                // Lzma reader owns its buffer rather than mutably borrowing it, so we have to drop
+                // it separately
+                if let Ok(mut remaining) = r.into_inner().finish() {
+                    let _ = copy(&mut remaining, &mut sink());
+                }
+                return;
+            }
+            #[cfg(feature = "legacy-zip")]
+            ZipFileReader::Implode(r) => {
+                // Lzma reader owns its buffer rather than mutably borrowing it, so we have to drop
+                // it separately
+                if let Ok(mut remaining) = r.into_inner().finish() {
+                    let _ = copy(&mut remaining, &mut sink());
+                }
+                return;
+            }
```

Comment:
These branches can be combined.
```suggestion
            #[cfg(feature = "legacy-zip")]
            ZipFileReader::Shrink(r) | ZipFileReader::Reduce(r) | ZipFileReader::Implode(r) => {
                // Lzma reader owns its buffer rather than mutably borrowing it, so we have to drop
                // it separately
                if let Ok(mut remaining) = r.into_inner().finish() {
                    let _ = copy(&mut remaining, &mut sink());
                }
                return;
            }
```

Reply:
type differs so it doesn't work to merge it.